app/components/graph-editor/drawable-interfaces.ts

```diff
@@ -0,0 +1,398 @@
+     *   is checked against the given source and destination nodes.
+     * 
+     *   This method must be called before calling `createEdge`.
+     */
+    canCreateEdge(
+        src: DrawableNode,
+        dst: DrawableNode,
+        like?: DrawableEdge
+    ): boolean;
+
+    /**
+     * createEdge  
+     *   Creates a drawable edge with a source and destination node.
+     * 
+     *   If `like` is specified, a drawable edge with a matching type of `like` is
+     *   created.
+     * 
+     *   The `canCreateEdge` method must be called to check if creating the edge
+     *   is valid.
+     */
+    createEdge(
+        src: DrawableNode,
+        dst: DrawableNode,
+        like?: DrawableEdge
+    ): DrawableEdge;
+
+    /**
+     * replaceEdge  
+     *   This method has been replaced by `createEdge` and is no longer needed.
+     */
+    replaceEdge(original: DrawableEdge, replacement: DrawableEdge): void;
+
+    /**
+     * removeEdge  
+     *   Guarantees that the given edge is not present in the graph.
+     */
+    removeEdge(edge: DrawableEdge): void;
+
+}
+
+/**
+ * DrawableEdge  
+ *   Interface that exposes drawable edge properties.
+ */
+export interface DrawableEdge {
+
+    /**
+     * source  
+     *   The source node of the edge.
+     */
+    source: DrawableNode | null;
+
+    /**
+     * destination  
+     *   The destination node of the edge.
+     */
+    destination: DrawableNode | null;
+
+    /**
+     * showSourceArrow  
+     *   True to draw an arrow pointing to the source node; otherwise, false.
+     */
+    showSourceArrow: boolean;
+
+    /**
+     * showDestinationArrow  
+     *   True to draw an arrow pointing to the destination node; otherwise, false.
+     */
+    showDestinationArrow: boolean;
+
+    /**
+     * color  
+     *   The color of the edge. This can be any valid `CSS` color string.
+     */
+    color: string;
+
+    /**
+     * lineStyle  
+     *   The line style of the edge. This can be `solid`, `dotted`, or `dashed`.
+     */
+    lineStyle: string;
+
+    /**
+     * lineWidth  
+     *   The width of the edge. This value must be non-negative.
+     */
+    lineWidth: number;
+
+    /**
+     * label  
+     *   The text label to be displayed by the edge.
+     */
+    label: string;
+
+    // TODO:
+    // more display properties.
+}
+
+/**
+ * DrawableNode  
+ *   Interface that exposes drawable node properties.
+ */
+export interface DrawableNode {
+
+    /**
+     * x  
+     *   The x-coordinate of the center of the node.
+     */
+    x: number;
+
+    /**
+     * y  
+     *   The y-coordinate of the center of the node.
+     */
+    y: number;
+
+    /**
+     * label  
+     *   The text label to be displayed in the node.
+     */
+    label: string;
+
+    /**
+     * shape  
+     *   The shape of the node.
+     */
+    shape: string; // TODO: for now this only supports circles and squares.
```

Comment:
I am also aware of this as well. I was thinking of doing something along the lines of union string types; e.g. `"circle" | "square" | "image"`, I just haven't gotten around to it.